README.md

```diff
@@ -14,16 +14,12 @@
   - [Badge](#badge)
 - [Definitions](#definitions)
 - [API](#api)
-  - [IPLD format utils](#ipld-format-utils)
-    - [`util.serialize(dagNode, callback)`](#utilserializedagnode-callback)
-    - [`util.deserialize(binaryBlob, callback)`](#utildeserializebinaryblob-callback)
-    - [`util.cid(binaryBlob[, options], callback)`](#utilcidbinaryblob-options-callback)
-  - [Local resolver methods](#local-resolver-methods)
-    - [`resolver.resolve(binaryBlob, path, callback)`](#resolverresolvebinaryblob-path-callback)
-    - [`resolver.tree(binaryBlob, callback)`](#resolvertreebinaryblob-callback)
```

Comment:
Where did these methods go?

Reply:
I would like to see the new `js-ipld` with the formats updated and resolving through things with all of these updates. It seems that we are preemptively removing things but I might be missing something.

Reply:
From the commit message:

> The biggest change is that there's not `resolve` API anymore. From
> now on you access the properties of the JavaScript objects directly.

So instead of `resolve.resolve(binaryBlob, '/deeply/nested/path)' you deserialize  the node first (which resolve also needs to do) and then on the resolved node, let's cal it `ipldNode` you do `ipldNode.deeply.nested.path`.

For `tree()`-like functionality (if needed) you do what [dag-json is doing](https://github.com/mikeal/dag-json/blob/3ba08ee1ac92341038585d07f482dab7caf621e5/index.js#L109-L125).

Reply:
So now it is the user that needs to convert an IPLD Path into JS notation?

Reply:
Note that dag-pb supported resolve as in `cid/name-of-the-link-to-the-other-node` and `cid/links/0/hash/`. It was basically packing a transformation inside the format.

Reply:
One benefit of not implementing path resolution in the codec interface is that it actually **forces** parity between the in-memory object representation and the path representation. 

In the old interface, because it was a different method, it was possible (and I think this actually happened) where the path resolution was different from the de-serialized object representation.

Reply:
@daviddias I implemented this spec for dag-pb, also the named links. See https://github.com/ipld/js-ipld-dag-pb/pull/105 and especially https://github.com/ipld/js-ipld-dag-pb/pull/105/commits/f98f09d0e3d1a51aa4c367ecbcb035d822385881 to see how named links would work then.

Reply:
A few things about that PR:

In that change named links aren't added to the object if they match existing property names (`size`, `data`, etc) so not all property traversals will work.

If you're doing a traversal in the style of `resolver.resolve(binaryBlob, '/deeply/nested/path')` you have the option of throwing away everything you've traversed so far, whereas if you do `ipldNode.deeply.nested.path` you have to create the whole object in memory first which might not be great depending on your dataset.

Reply:
> That's the same behaviour as the current implementation

Hmm, could we make it better?

> You could implement the same semantics with JavaScript Proxies if needed.

What's the browser support for that like?  I found [babel-plugin-proxy](https://www.npmjs.com/package/babel-plugin-proxy) but the README says it's got really bad performance and don't recommend using it in production environments.

Reply:
> Hmm, could we make it better?

I would get rid of the named links traversal (I call it "IPFS traversal") or put it under a field (e.g. `ipfs`) so it won't collide. Though I think being just as broken as we are currently is good enough for now, especially since this is again dag-pb specific and shouldn't be a problem with other formats.

> What's the browser support for that like?

Pretty good: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy#Browser_compatibility

Reply:
> I would get rid of the named links traversal 

This is not a crazy idea.  It's only of limited for IPFS use because there isn't a one-to-one mapping between IPFS paths and IPLD paths for HAMT shards.

E.g.

```javascript
resolve.resolve(binaryBlob, '/deeply/nested/path.txt')
```

Might end up needing to be

```javascript
resolve.resolve(binaryBlob, '/deeply/A0/01/FC-nested/path.txt')
```

..and you don't know what the prefix for each segment is going to be until you descend into that level of the graph as it's dependent on the siblings.

Reply:
I mostly kept the named links to show that such things are still possible.

Reply:
I personally don't think that move to JS based traversal has following disadvantages:

1. As @achingbrain  pointed out that implies that that whole node is de-serialized in one step. Sure JS proxies could be utilized but they are very slow and are really difficult to get right as in ensure that they you don't end up creating memory leaks (we used them a lot for cross compartment boundaries in gecko & in my opinion for non trivial cases e.g `process.env` they're not worth it).

   I have being wanted to implement ipld resolver for [flatbuffers](https://google.github.io/flatbuffers/) who's primary advantage over proto-buffs is laziness - only relevant fields can be decoded without instantiating objects for intermediaries. It seems like with this move taking advantage of this would be impossible.

   Proxies only work if you can decode everything synchronously, if that is not the case then you wind up with bunch of promises you have to await on to get to the nested data, which IMO is less convenient than it is to provide a path.